allow calling `goToDef` on modifiers of named declarations

iisaduan opened the pull request:
Fixes #60308.

There are some symbols where goToDef/goToType previously did not return a definition. For those cases, I have not changed the behavior, this PR just makes it so that calling `goTo(Type)Definition` on a modifier will do the same behavior as if the call was triggered on the name.

The first commit refactors some existing `goToDefinition` keyword cases. It may be easier to review by looking at the first commit, and then the rest of the commits without the first one.

mjbvz reviewed (approved):
Thanks for looking into this!

DanielRosenwasser commented on src/services/goToDefinition.ts:
```diff
@@ -159,16 +157,30 @@ export function getDefinitionAtPosition(program: Program, sourceFile: SourceFile
+    switch (node.kind) {
+        case SyntaxKind.ReturnKeyword:
+            findFunctionDecl = n => {
+                return isClassStaticBlockDeclaration(n)
```
This feels like it's very much a special-case. Are you trying to just not break some sort of test case?

iisaduan replied:
It was existing code, so I'm not sure if there are more cases that would be an issue, but it breaks https://github.com/microsoft/TypeScript/blob/main/tests/cases/fourslash/goToDefinitionReturn5.ts

DanielRosenwasser replied:
It's probably just fine to update the baseline there.

DanielRosenwasser commented on src/services/goToDefinition.ts:
```diff
@@ -159,16 +157,30 @@ export function getDefinitionAtPosition(program: Program, sourceFile: SourceFile
-        const functionDeclaration = findAncestor(node, n => isFunctionLikeDeclaration(n));
-        const isGeneratorFunction = functionDeclaration && functionDeclaration.asteriskToken;
-        return isGeneratorFunction ? [createDefinitionFromSignatureDeclaration(typeChecker, functionDeclaration)] : undefined;
+    if (findFunctionDecl) {
+        const functionDeclaration = findAncestor(node, findFunctionDecl) as FunctionLikeDeclaration | undefined;
+        const isCorrectDeclaration = functionDeclaration && (!checkFunctionDeclaration || checkFunctionDeclaration(functionDeclaration));
+        return isCorrectDeclaration ? [createDefinitionFromSignatureDeclaration(typeChecker, functionDeclaration)] : undefined;
```
I kind of find the indirection here hard to follow. I can see wanting to find some common logic, but I'd suggest just writing it out explicitly in each branch for now.

DanielRosenwasser replied:
Also, the idea that you *wouldn't* want to still jump to the nearest function even if it's not `async` or a generator is something I'm not so sure of. So the idea that you need a `checkFunctionDeclaration` to validate that you got a function of the right kind feels like it would break down unnecessarily during refactoring.

iisaduan commented:
@mjbvz I addressed @DanielRosenwasser 's comments and added some more keywords to return definitions. Would you both take a look to see if the new behavior is reasonable?

mjbvz reviewed (approved):
Thanks @iisaduan! Just tested this in VS Code and the new behavior looks great. Thank you for looking into this

DanielRosenwasser commented:
@typescript-bot pack this
@typescript-bot test tsserver top300
